apps/remote_control/lib/lexical/remote_control/search/store/backends/ets/wal.ex

```diff
@@ -0,0 +1,424 @@
+          wal
+          | update_log: log,
+            update_log_name: wal_name,
+            checkpoint_version: checkpoint_version
+        }
+
+        {:ok, new_wal}
+
+      error ->
+        error
+    end
+  end
+
+  defp update_wal_name(%__MODULE__{} = wal) do
+    :"updates_for_#{Project.name(wal.project)}_v#{wal.schema_version}"
+  end
+
+  # Updates
+  defp apply_updates(%__MODULE__{} = wal) do
+    stream_updates(wal, wal.update_log, :start)
+  end
+
+  defp stream_updates(%__MODULE__{} = wal, log, continuation) do
+    case :disk_log.chunk(log, continuation, @chunk_size) do
+      {continuation, items} when is_list(items) ->
+        apply_relevant_items(wal, items)
+        stream_updates(wal, log, continuation)
+
+      {continuation, items, _bad_bytes} ->
+        apply_relevant_items(wal, items)
+        stream_updates(wal, log, continuation)
+
+      :eof ->
+        :ok
+
+      {:error, _} = error ->
+        error
+    end
+  end
+
+  defp apply_relevant_items(%__MODULE__{} = wal, items) do
+    checkpoint_version = wal.checkpoint_version
+
+    items
+    |> Stream.filter(fn operation(id: id) -> id >= checkpoint_version end)
+    |> Enum.each(fn operation(function: function, args: args) ->
+      apply(:ets, function, args)
+    end)
+  end
+
+  defp get_wal_operations(%__MODULE__{} = wal) do
+    stats = :disk_log.info(wal.update_log)
+    Keyword.get(stats, :items, 0)
+  end
+
+  # Checkpoints
+  defp needs_checkpoint?(%__MODULE__{} = wal) do
+    get_wal_operations(wal) >= wal.max_wal_operations
+  end
+
+  defp maybe_checkpoint(%__MODULE__{} = wal) do
+    with true <- needs_checkpoint?(wal),
+         {:ok, new_wal} <- checkpoint(wal) do
+      {:ok, new_wal}
+    else
+      _ ->
+        {:ok, wal}
+    end
+  end
+
+  defp do_checkpoint(%__MODULE__{} = wal) do
+    checkpoint_version = Identifier.next_global!()
+    checkpoint_file_name = checkpoint_file_name(checkpoint_version)
+
+    log_path = wal |> wal_directory() |> Path.join(checkpoint_file_name)
+    log_name = checkpoint_log_name(wal.project)
+
+    with {:ok, log} <- :disk_log.open(name: log_name, file: String.to_charlist(log_path)),
+         :ok <- checkpoint_ets_table(wal, log),
+         :ok <- :disk_log.close(log),
+         :ok <- :disk_log.truncate(wal.update_log) do
+      new_wal = %__MODULE__{wal | checkpoint_version: checkpoint_version}
+      delete_old_checkpoints(new_wal)
+      {:ok, new_wal}
+    else
+      error ->
+        # Checkpoint loading failed. Give up and start over
+        delete_old_checkpoints(wal)
+        error
+    end
+  end
+
+  defp checkpoint_ets_table(%__MODULE__{} = wal, log) do
+    log_chunks = fn
+      item, {@chunk_size, items} ->
+        :disk_log.log_terms(log, Enum.reverse(items))
+        {1, [item]}
+
+      item, {count, items} ->
+        {count + 1, [item | items]}
+    end
+
+    {_count, items} = :ets.foldl(log_chunks, {0, []}, wal.ets_table)
+    :disk_log.log_terms(log, items)
```

Comment:
Do items need to be reversed here as well?

```suggestion
    :disk_log.log_terms(log, Enum.reverse(items))
```